static plotting for spatial data

giovp opened the pull request:
**IMPORTANT: Please search among the [Pull requests](../pulls) before creating one.**

## Description
Add :func:`squidpy.pl.spatial_scatter` and :func:`squidpy.pl.spatial_segment` to statically plot spatial omics data.


## TODO
- [x] add segmentation plotting capabilities #427 
- [x] add some type of behaviour for alpha in continuous values. The reason for this is that if with umap you don't care to see points where the value is e.g. 0 (and you can just subset adata), in spatial you do cause you still want to see the full tissue (and this doesn't happen by proxy if you subset adata, cause it also crop the tissue). Not super sure how to handle it yet.
- [x] saving
- [x] docs
- [x] tests

closes #242 #247 #444

giovp commented:
Ok here's a demo on what's currently doing:

```python
import squidpy as sq

adata1 = sq.datasets.visium_hne_adata()
adata2 = adata1.copy() # copy adata
adata2.uns["spatial"] = {} # create fake new library_id
adata2.uns["spatial"]["V2_Adult_Mouse_Brain"] = adata1.uns["spatial"][
    "V1_Adult_Mouse_Brain"
]

# concat adatas putting library
adata_temp = ad.concat(
    {"V1_Adult_Mouse_Brain": adata1, "V2_Adult_Mouse_Brain": adata2},
    label="batch_key",
    uns_merge="only",
)

from squidpy.pl._spatial import spatial

spatial(
    adata_temp,
    batch_key="batch_key",
    cmap="inferno",
    shape="hex",
    size=[1.2, 1.0],
    color=["Sox17", "cluster"],
    add_outline=True,
    outline_width=(0.05, 0.05),
    crop_coord=[[1500, 2800, 1500, 2800], [1500, 2800, 1500, 2800]],
    scalebar_dx=2.0,
    scalebar_kwargs={"scale_loc": "bottom", "location": "lower right"},
)

```
![image](https://user-images.githubusercontent.com/25887487/148434979-5e365f90-8b3b-4ce9-b0a9-b10ab6f900b1.png)

I followed `scanpy.pl.embedding` module and adapted on the way. The main additions are:
- it can now plot multiple library_ids (different visium slides) saved in the same adata. It essentially treats `library_id` the same way `sc.pl.embedding` treats `components`. This is very useful for storing everything in a single adata object. 
- it can plot with different shapes (added squares and hexagons), useful for e.g. dbitseq.
- it can add the scale bar (important feature request).

code is messy and I'm still not super sure on the behaviour of the image-related arguments (e.g. spot size, scale factors, scalebars etc.). For now, they either have to be a single instance (e.g. a float, a str), so that that value is propagate to all slides, or a list that matches the `library_id` length. I think is sensible but it's also debatable. Would love to get feedback.

things that are still missing:
- [ ] add segmentation plotting capabilities #427 
- [ ] add some type of behaviour for alpha in continuous values. The reason for this is that if with umap you don't care to see points where the value is e.g. 0 (and you can just subset adata), in spatial you do cause you still want to see the full tissue (and this doesn't happen by proxy if you subset adata, cause it also crop the tissue). Not super sure how to handle it yet.

plus ofc docs, tests etc. I'll take another look tomorrow and then request for first round of review.

giovp commented:
Ok I think this a good stage to start testing out behaviour and overall structure. Things I'd like to get feedback on:
- **code structure**: it follows `sc.pl.embedding` which can seem a bit messy (it's a big for loop) but intuitive to follow(?).
- **arguments**: again, it follows `sc.pl.embedding`  but again I think it's intuitive cause people are used to arguments. They are many though and some could be grouped into `kwargs`, although then we'd have to find a way to explict some of them in docs in order to highlight.
- **anything else**

Here's 2 code snippets to get started

### Visium

<details>
<summary>Details</summary>

```python
import squidpy as sq

adata1 = sq.datasets.visium_hne_adata()
adata2 = adata1.copy() # copy adata
adata2.uns["spatial"] = {} # create fake new library_id
adata2.uns["spatial"]["V2_Adult_Mouse_Brain"] = adata1.uns["spatial"][
    "V1_Adult_Mouse_Brain"
]

# concat adatas putting library
adata_temp = ad.concat(
    {"V1_Adult_Mouse_Brain": adata1, "V2_Adult_Mouse_Brain": adata2},
    label="batch_key",
    uns_merge="only",
	pairwise=True,
)

from squidpy.pl._spatial import spatial

spatial(
    adata_temp,
    batch_key="batch_key",
    cmap="inferno",
    shape="circle",
    size=[0.3, 0.5],
    color=["Sox17", "cluster"],
    edges=True,
    edges_width=5,
    add_outline=True,
    outline_width=(0.05, 0.05),
    crop_coord=[[1500, 2800, 1500, 2800], [1500, 2800, 1500, 2800]],
    scalebar_dx=2.0,
    scalebar_kwargs={"scale_loc": "bottom", "location": "lower right"},
)
```

![image](https://user-images.githubusercontent.com/25887487/148583087-57ac0f93-b68f-4d83-92dc-b59dd7a581c8.png)


</details>

### Seqfish

<details>
<summary>Details</summary>

```python
import squidpy as sq

adata_seqfish1 = sq.datasets.seqfish()
sq.gr.spatial_neighbors(adata_seqfish1)
adata_seqfish2 = adata_seqfish1.copy()

adata_seqfish = ad.concat(
    {"segfish1": adata_seqfish1, "seqfish2": adata_seqfish2},
    label="batch_key",
    uns_merge="only",
    pairwise=True,
)

from squidpy.pl._spatial import spatial

spatial(
    adata_seqfish,
    batch_key="batch_key",
    shape=None,
    cmap="inferno",
    edges=True,
    edges_width=2,
    #     shape="hex",
    size=[30, 50],
    color=["Wnt8a", "celltype_mapped_refined"],
    add_outline=True,
    outline_width=(0.2, 0.2),
    crop_coord=[[1, 2, 1, 2], [1, 2, 1, 2]],
    scalebar_dx=2.0,
    scalebar_kwargs={"scale_loc": "bottom", "location": "lower right"},
)
```

![image](https://user-images.githubusercontent.com/25887487/148583532-66860132-9381-4481-a4e1-091995f3576d.png)

</details>

michalk8 commented on squidpy/pl/_spatial.py:
```diff
@@ -0,0 +1,764 @@
+            f"Specified `size_key: {size_key}` does not exist and size is `None`,\
+            available keys are: `{scalefactors}`.\n Specify valid `size_key` or `size`."
+        )
+    if size is None:
```
Why can size be `None`, just make it default `1.0`

giovp replied:
because in the case of non-visium data, if None it computes optimal one based on scanpy

```size = 120000 / adata.shape[0] if size is None else size```

michalk8 replied:
Ok, sounds good.

michalk8 reviewed (changes requested):
General remark w.r.t. to exception strings - please try to make it consistent with the rest of the package, e.g. ``Expected `<var_name>` to be of type `<type_name>`, found `{type(<var_name>).__name__}`.`` or
``Unable to find spatial data in `adata.obsm[{spatial_key}!r]`.``

giovp commented on squidpy/pl/_spatial.py:
```diff
@@ -0,0 +1,764 @@
+        return values, False
+    elif is_categorical_dtype(values):
+        # use scanpy _get_palette to set palette if not present
+        color_map = {k: to_hex(v) for k, v in _get_palette(adata, value_to_plot, palette).items()}
```
this is actually a point we should consider, since indeed we are subsetting and so can't do it the scanpy way. Shall we handle it at the beginning and just add a palette color if missing the way scanpy does? thoughts @hspitzer ?

giovp replied:
actually, you ofc were right, changed the way we handle it and the palette if not present is set upstream

giovp commented on squidpy/pl/_spatial.py:
```diff
@@ -0,0 +1,764 @@
+    if isinstance(norm, Normalize) or not isinstance(norm, Sequence):
+        norm = [norm]
+
+    return vmin, vmax, vcenter, norm
```
namedtuple are so coool! thanks for this pointer, used them both in input and output

giovp commented on squidpy/pl/_spatial.py:
```diff
@@ -0,0 +1,764 @@
+            )
+
+
+def _get_unique_map(dic: Mapping[str, Any]) -> Any:
```
I must say this is the only solution I could find in order to not make mypy fails badly, I've seen issues on how set are handled in mypy, so think not too bad to keep it.

giovp commented on squidpy/pl/_spatial.py:
```diff
@@ -0,0 +1,764 @@
+    return _crop_coord
+
+
+def _maybe_subset(adata: AnnData, batch_key: str | None = None, library_id: str | None = None) -> AnnData:
```
done

giovp commented on squidpy/pl/_spatial.py:
```diff
@@ -0,0 +1,764 @@
+    return _crop_coord
+
+
+def _maybe_subset(adata: AnnData, batch_key: str | None = None, library_id: str | None = None) -> AnnData:
```
great suggestion

giovp commented on squidpy/pl/_spatial.py:
```diff
@@ -0,0 +1,764 @@
+def _check_crop_coord(
+    crop_coord: Tuple[float, float, float, float],
+    scale_factor: float,
+) -> Tuple[float, ...]:
```
really great suggestion, done!

giovp commented on squidpy/pl/_spatial.py:
```diff
@@ -0,0 +1,764 @@
+    batch_key: Optional[str | None] = None,
+) -> Sequence[NDArrayA]:
+
+    coords = adata.obsm[spatial_key]
```
this is handled with `_assert_spatial_basis(adata, spatial_key)` above at the very beginning

giovp commented on squidpy/pl/_spatial.py:
```diff
@@ -0,0 +1,764 @@
+    coords = adata.obsm[spatial_key]
+    if batch_key is None:
+        data_points = [np.multiply(coords, sf) for sf in scale_factor]
+    else:
```
This is done in `_get_spatial_attrs` which is called above. But added a check.

giovp commented on squidpy/pl/_spatial.py:
```diff
@@ -0,0 +1,764 @@
+                from matplotlib_scalebar.scalebar import ScaleBar
+        except ImportError:
+            raise ImportError(
+                "To use scalebar functionality please install `matplotlib_scalebar` with: \n \
```
mmh no I think it's fine to have a relative import.

giovp commented on squidpy/pl/_spatial.py:
```diff
@@ -0,0 +1,764 @@
+    palette: Optional[Sequence[str] | str | Cycler | None] = None,
+    color_map: Union[Colormap, str, None] = None,
+    cmap: Union[Colormap, str, None] = None,
+    na_color: Optional[str | Tuple[float, ...] | None] = (0.0, 0.0, 0.0, 0.0),
```
think more intuitive this way, since in contrast to normal scatter, makes big difference to have it gray or not. Will remove none

giovp commented on squidpy/pl/_spatial.py:
```diff
@@ -0,0 +1,764 @@
+    wspace = 0.75 / rcParams["figure.figsize"][0] + 0.02 if wspace is None else wspace
+
+    # logic for image v. non-image data is handled here
+    if shape is not None:
```
made another helper function think clearer cause settings are different.

giovp commented on squidpy/pl/_spatial.py:
```diff
@@ -0,0 +1,764 @@
+    legend_fontweight: Union[int, _FontWeight] = "bold",
+    legend_loc: str = "right margin",
+    legend_fontoutline: Optional[int] = None,
+    na_in_legend: bool = True,
```
I'd keep it in for now

giovp commented on squidpy/pl/_spatial.py:
```diff
@@ -0,0 +1,764 @@
+        img_key = _get_unique_map(image_mapping)  # get intersection of image_mapping.values()
+        img_key = img_key[0]  # get first of set
+    else:
+        if img_key not in image_mapping.values():
```
mmh no image mapping already contains all possible values of image res with the passed library_ids.

giovp commented on squidpy/pl/_spatial.py:
```diff
@@ -0,0 +1,764 @@
+        )
+
+    if img_key is None:
+        img_key = _get_unique_map(image_mapping)  # get intersection of image_mapping.values()
```
as above, decided not to cause mypy fails

giovp commented on squidpy/_constants/_pkg_constants.py:
```diff
@@ -128,6 +124,68 @@ def spot_diameter(cls, adata: AnnData, spatial_key: str, library_id: Optional[st
+            library_id: Optional[Sequence[str] | str] = None,
+        ) -> Mapping[str, Sequence[str]]:
+            library_id = cls._haystack(adata, spatial_key, library_id, sub_key)
+            mapping = {i: list(adata.uns[spatial_key][i][sub_key].keys()) for i in library_id}
```
what do you mean?

michalk8 replied:
You can just `return {i: list(adata.uns[spatial_key][i][sub_key].keys()) for i in library_id}`

giovp commented on squidpy/pl/_spatial.py:
```diff
@@ -0,0 +1,764 @@
+    groups: Optional[Sequence[str] | str | None] = None,
+    use_raw: Optional[bool | None] = None,
+    layer: Optional[str | None] = None,
+    alt_var: Optional[str | None] = None,
```
cause we might not have genes but proteins, metabolytes and other so best to be neutral

giovp commented on squidpy/_constants/_pkg_constants.py:
```diff
@@ -128,6 +124,68 @@ def spot_diameter(cls, adata: AnnData, spatial_key: str, library_id: Optional[st
+                if not any(i in library_id for i in haystack):
+                    raise KeyError(f"`library_id`: {library_id}` not found in `{sorted(haystack)}`.")
+                if sub_key is not None:
+                    if not all(sub_key in i for i in [adata.uns["spatial"][i].keys() for i in library_id]):
```
I modifed with `not all` above in order to fail early

giovp commented on squidpy/pl/_spatial.py:
```diff
@@ -0,0 +1,764 @@
+    layer: Optional[str | None] = None,
+    alt_var: Optional[str | None] = None,
+    projection: Literal["2d", "3d"] = "2d",
+    edges: bool = False,
```
also patchcollection uses edges though, would keep

giovp commented on squidpy/pl/_spatial.py:
```diff
@@ -0,0 +1,764 @@
+    crop_coord: Optional[Sequence[Tuple[int, int, int, int]] | Tuple[int, int, int, int] | None] = None,
+    bw: bool = False,
+    alpha_img: float = 1.0,
+    color: Optional[Sequence[str | None] | str | None] = None,
```
nice, done

giovp commented on squidpy/pl/_spatial.py:
```diff
@@ -0,0 +1,764 @@
+            )
+            cax = ax.add_collection(_cax)
+
+        ax.set_yticks([])
```
done

giovp commented on squidpy/pl/_spatial.py:
```diff
@@ -0,0 +1,764 @@
+
+    # make plots
+    library_idx = range(len(_library_id))
+    for count, (value_to_plot, _lib_count) in enumerate(itertools.product(color, library_idx)):
```
extracted the ax_decorate

giovp commented on squidpy/pl/_spatial.py:
```diff
@@ -0,0 +1,764 @@
+    kwargs["cmap"] = cmap
+
+    # set cmap_img
+    cmap_img = "gray" if bw else None
```
nice, done

giovp commented on squidpy/pl/_spatial.py:
```diff
@@ -0,0 +1,316 @@
+    norm: _Normalize = None,
+    na_color: str | Tuple[float, ...] = (0, 0, 0, 0),
+    # image
+    img: _SeqArray | bool = None,
```
so here and also in `seg` I've implemented type of argument like this in order to use None. Let me explain, I basically want this argument to store both logic for using it as well as its values.

So if the arg is True or array then it's used, if it's False it's not used. It's similar to what we tried to do here:  https://github.com/theislab/scanpy/blob/8736fc3b02a6af3b7214481c2cf78a3f8f5e06bd/scanpy/plotting/_tools/scatterplots.py#L1371

an alternative way that I can come up with is to add an additional argument e.g. `plot_img` which would then replace it in `_set_spatial_attrs` although I'm against adding yet another parameter. Would be curious to hear your thoughts.

giovp commented on squidpy/pl/_spatial.py:
```diff
@@ -0,0 +1,316 @@
+    img_cmap: Colormap | str | None = None,
+    img_channel: int | None = None,
+    # segment
+    seg: _SeqArray | bool = None,
```
here same point as above re `img`

giovp commented on squidpy/pl/_spatial_utils.py:
```diff
@@ -0,0 +1,875 @@
+    )
+    if seg_boundaries:
+        seg_bound: NDArrayA = np.clip(seg_im - find_boundaries(val_im)[:, :, None], 0, 1)
+        seg_bound = np.dstack((seg_bound, np.where(val_im > 0, 1, 0)))  # add transparency here
```
this is to add transparency, which wouldn't work out of the box with `label2rgb`

giovp commented on squidpy/pl/_spatial_utils.py:
```diff
@@ -0,0 +1,875 @@
+    seg_transparent: bool = False,
+) -> NDArrayA:
+
+    cell_id = np.array(cell_id)
```
this is to avoid that cell_id is modifed in place across iterations.

giovp commented on squidpy/pl/_spatial_utils.py:
```diff
@@ -0,0 +1,875 @@
+        val_im: NDArrayA = map_array(seg, cell_id, color_vector.codes + 1)  # type: ignore
+        cols = colors.to_rgba_array(color_vector.categories)  # type: ignore
+    else:
+        val_im = map_array(seg, cell_id, cell_id)
```
this is needed to "filter out" non-existings segmentation masks and then replace with color in the `label2rgb` call

giovp reviewed (commented):
hi @michalk8 I've done a big refactoring and hope now it's clearer. I've incorporated basically every comment you had except some which were more related to styling anyway. Would appreciate another round of review, thanks a lot in advance

ivirshup reviewed (commented):
Looking good! Just a bit of feedback around segmentation plotting code.

giovp commented on squidpy/pl/_spatial_utils.py:
```diff
@@ -0,0 +1,875 @@
+    return ax, cax
+
+
+def _plot_segment(
```
makes sense, imported _map_seg_color there

giovp commented on squidpy/pl/_spatial.py:
```diff
@@ -0,0 +1,314 @@
+        else:
+            ax, cax = _plot_segment(
+                img=_seg, ax=ax, cmap_params=cmap_params, color_params=color_params, categorical=categorical, **kwargs
+            )
```
yeah from comment above now looks better

giovp commented on squidpy/pl/_spatial_utils.py:
```diff
@@ -0,0 +1,875 @@
+    cmap_params: CmapParams,
+    seg_erosionpx: int | None = None,
+    seg_boundaries: bool = False,
+    seg_transparent: bool = False,
```
I see the point, I could actually infer it from the na_color indeed. Will do that and then clarify in docs

giovp commented on squidpy/pl/_spatial_utils.py:
```diff
@@ -0,0 +1,875 @@
+
+    _cax = ax.imshow(
+        img,
+        rasterized=sc_settings._vector_friendly,
```
good point, changed

michalk8 commented on squidpy/pl/_spatial_utils.py:
```diff
@@ -0,0 +1,897 @@
+                library_id = adata.obs[library_key].cat.categories.tolist()
+            _assert_value_in_obs(adata, key=library_key, val=library_id)
+            return library_id
+        else:
```
Unnecessary `else`.

michalk8 commented on squidpy/pl/_spatial_utils.py:
```diff
@@ -0,0 +1,897 @@
+        scalefactor_mapping = Key.uns.library_mapping(adata, spatial_key, Key.uns.scalefactor_key, library_id)
+        scalefactors = _get_unique_map(scalefactor_mapping)
+    except KeyError:
+        scalefactors = None
```
I'd add `logging.debug(...)` just to make it easier in the future to hunt for bugs.

michalk8 commented on squidpy/pl/_spatial_utils.py:
```diff
@@ -0,0 +1,897 @@
+
+        if size_key not in scalefactors and size is None:
+            raise ValueError(
+                f"Specified `size_key: {size_key}` does not exist and size is `None`,\
```
Remove `\`, the message will be ugly.

michalk8 commented on squidpy/pl/_spatial_utils.py:
```diff
@@ -0,0 +1,897 @@
+
+def _subs(adata: AnnData, library_key: str | None = None, library_id: str | None = None) -> AnnData:
+    try:
+        return adata[adata.obs[library_key] == library_id].copy()
```
Is the copy necessary? I guess it is, just asking, can be problematic making a lots of hidden copies just for plotting.

giovp replied:
yeah it is, do you think it'd clutter the memory session? a solution for that?

michalk8 replied:
I've look at the place where you call `_subs`, some suggestions:
- you call it twice (once when calling `_set_color_source_vec`, other time in `_plot_edges`), should be called only once
- as for copying, the problem mostly lies with `AnnData` when you tried to write to a view (it creates a copy for you)
  looking at `_set_color_source_vec`, you don't write anything to `adata`, so it should be fine
  as for `_plot_edges`, it's the same story
  
so I think the copy can be removed (unless I'm missing something).

giovp replied:
good point, removed

michalk8 commented on squidpy/pl/_spatial_utils.py:
```diff
@@ -0,0 +1,897 @@
+    return sorted(set.intersection(*map(set, dic.values())))
+
+
+def _get_list(var: Any, _type: Any, ref_len: int | None = None) -> List[Any] | Any:
```
`_type: Type[Any]`

ivirshup replied:
```suggestion
def _get_list(var: Any, _type: Type[Any], ref_len: int | None = None) -> List[Any]:
```

michalk8 commented on squidpy/pl/_spatial_utils.py:
```diff
@@ -0,0 +1,897 @@
+        if ref_len is None:
+            return [var]
+        return [var] * ref_len
+    elif isinstance(var, list):
```
Minor pet peeve, no need for `elif`.

michalk8 commented on squidpy/pl/_spatial_utils.py:
```diff
@@ -0,0 +1,897 @@
+        if (ref_len is not None) and (ref_len != len(var)):
+            raise ValueError(f"Var len: {len(var)} is not equal to ref len: {ref_len}. Please Check.")
+        return var
+    else:
```
Minor pet peeve, no need for `else`, feel free to ignore.

michalk8 commented on squidpy/pl/_spatial_utils.py:
```diff
@@ -0,0 +1,897 @@
+    groups: _SeqStr = None,
+    palette: Palette_t = None,
+    na_color: str | Tuple[float, ...] | None = None,
+) -> Tuple[NDArrayA | pd.Series, NDArrayA, bool]:
```
`Tuple[NDArrayA | pd.Series, NDArrayA, bool] -> Tuple[NDArrayA | pd.Series | None, NDArrayA, bool]`

michalk8 commented on squidpy/pl/_spatial_utils.py:
```diff
@@ -0,0 +1,897 @@
+        return None, np.full(to_hex(na_color), adata.n_obs), False
+    if not is_categorical_dtype(color_source_vector):
+        return None, color_source_vector, False
+    else:
```
Minor pet peeve, no need for `else`, feel free to ignore.

michalk8 commented on squidpy/_constants/_constants.py:
```diff
@@ -105,3 +105,10 @@ class RipleyStat(ModeEnum):
+
+
+@unique
+class ScatterShape(ModeEnum):
```
`class ScatterShape(ModeEnum) ->  class ScatterShape(str, ModeEnum)`

giovp replied:
it doesn't actually work

michalk8 replied:
In what sense? Running
```python
class X(str, sq._constants._constants.ModeEnum):
        foo = 'foo'
assert isinstance(X.foo, str)  # passed
```

michalk8 commented on squidpy/pl/_spatial_utils.py:
```diff
@@ -0,0 +1,897 @@
+    y: NDArrayA,
+    s: float,
+    c: NDArrayA,
+    shape: _AvailShapes | None = ScatterShape.CIRCLE.s,  # type: ignore[assignment]
```
If `ScatterShape` is subclass of `str`, remove `.s`.

michalk8 commented on squidpy/pl/_spatial_utils.py:
```diff
@@ -0,0 +1,897 @@
+    shape = ScatterShape(shape)  # type: ignore[assignment]
+    if TYPE_CHECKING:
+        assert isinstance(shape, ScatterShape)
+    shape = shape.s
```
Use `shape = ScatterShape(shape)` to allow only available shapes.

michalk8 commented on squidpy/pl/_spatial_utils.py:
```diff
@@ -0,0 +1,897 @@
+        assert isinstance(shape, ScatterShape)
+    shape = shape.s
+
+    if shape == ScatterShape.CIRCLE.s:
```
Change to `shape == ScatterShape.CIRCLE`

giovp replied:
inheritance as suggested above doesn't work

michalk8 commented on squidpy/pl/_spatial_utils.py:
```diff
@@ -0,0 +1,897 @@
+    edges_width: float = 0.1,
+    edges_color: str | Sequence[float] | Sequence[str] = "grey",
+    connectivity_key: str = Key.obsp.spatial_conn(),
+    edges_kwargs: Mapping[str, Any] = MappingProxyType({}),
```
Any reason to use `edges_kwargs` instead of `**kwargs`?

giovp replied:
no, good point

michalk8 commented on squidpy/pl/_spatial_utils.py:
```diff
@@ -0,0 +1,897 @@
+
+    # handle title
+    if title is not None:
+        if isinstance(title, list) and len(title) != n_plots:
```
Would use `isinstance(title, (tuple, list))` just in case, but no strong opinion.

michalk8 commented on squidpy/pl/_spatial_utils.py:
```diff
@@ -0,0 +1,897 @@
+        ax.set_aspect("equal")
+        ax.invert_yaxis()
+
+    if crops is not None:
```
Call `crops.to_tuple` only once and save it to some variable.

michalk8 commented on squidpy/pl/_spatial_utils.py:
```diff
@@ -0,0 +1,897 @@
+
+    # set palette if missing
+    for c in color:
+        if c in adata.obs.columns and is_categorical_dtype(adata.obs[c]) and c is not None:
```
I guess `c is not None` should be first in the `if` or completely removed, since 99.999% it will not be a column in `.obs`

michalk8 commented on squidpy/pl/_spatial_utils.py:
```diff
@@ -0,0 +1,897 @@
+        use_raw = layer is None and adata.raw is not None
+    if use_raw and layer is not None:
+        raise ValueError(
+            "Cannot use both a layer and the raw representation. Was passed:" f"use_raw={use_raw}, layer={layer}."
```
Remove the `black` artifact `:" f"`.

michalk8 commented on squidpy/pl/_spatial_utils.py:
```diff
@@ -0,0 +1,897 @@
+    # logic for image v. non-image data is handled here
+    shape = ScatterShape(shape).s if shape is not None else shape  # type: ignore
+    if TYPE_CHECKING:
+        assert isinstance(shape, ScatterShape)
```
This is not correct, `shape` can also be `None`

michalk8 commented on squidpy/pl/_spatial_utils.py:
```diff
@@ -0,0 +1,897 @@
+        raise ValueError(f"`use_raw={use_raw}` but AnnData object does not have raw.")
+
+    # logic for image v. non-image data is handled here
+    shape = ScatterShape(shape).s if shape is not None else shape  # type: ignore
```
`.s` will not be necessary once `ScatterShape` also subclasses string.

michalk8 reviewed (changes requested):
Looks nice, just some minor comments

ivirshup commented on squidpy/pl/_spatial_utils.py:
```diff
@@ -0,0 +1,897 @@
+    img = _map_color_seg(
+        seg=seg,
+        cell_id=cell_id,
+        color_vector=color_vector,
+        color_source_vector=color_source_vector,
+        cmap_params=cmap_params,
+        seg_erosionpx=seg_erosionpx,
+        seg_boundaries=seg_boundaries,
+        na_color=na_color,
+    )
+
+    _cax = ax.imshow(
+        img,
+        rasterized=True,
+        cmap=cmap_params.cmap if not categorical else None,
+        norm=cmap_params.norm if not categorical else None,
```
Why do both `_map_color_seg` and `imshow` get `cmap` and `norm`? Shouldn't you not need these for `imshow` since you're directly passing it an RGBA array?

giovp replied:
it's needed for the colorbar and the legend.

giovp commented on squidpy/_constants/_pkg_constants.py:
```diff
@@ -128,6 +132,63 @@ def spot_diameter(cls, adata: AnnData, spatial_key: str, library_id: Optional[st
+                if sub_key is not None:
+                    if not all(sub_key in lib for lib in [adata.uns[spatial_key][lib].keys() for lib in library_id]):
+                        raise KeyError(
+                            f"`{sub_key}` not found in `adata.uns[{spatial_key}]['library_id'].keys()`\
```
good point, will pay attention

giovp commented on squidpy/pl/_graph.py:
```diff
@@ -53,17 +53,34 @@ def _get_data(adata: AnnData, cluster_key: str, func_name: str, **kwargs: Any) -
+        len_cat = adata.obs[cluster_key].cat.categories.shape[0]
+        if palette in plt.colormaps():
+            cmap = plt.get_cmap(palette)
+            _palette = [to_hex(x) for x in cmap(np.linspace(0, 1, len_cat))]
```
great catch

giovp commented on squidpy/pl/_graph.py:
```diff
@@ -29,7 +29,7 @@
+Palette_t = Optional[Union[str, ListedColormap, None]]
 
 
 def _maybe_set_colors(source: AnnData, target: AnnData, key: str, palette: str | None = None) -> None:
```
yeah there is a bunch of this utils in `pl._graph` that should be moved in a util module. Not sure if for this PR?

michalk8 replied:
Yes, `_maybe_set_colors` and `_get_palette` in `pl._graph` should also be moved to `_color_utils`. I'd preferably move this (+ the other color functions in this PR), but if you don't want to, let's just keep it in mind and do it another time.

giovp replied:
done

giovp commented on squidpy/pl/_graph.py:
```diff
@@ -53,17 +53,34 @@ def _get_data(adata: AnnData, cluster_key: str, func_name: str, **kwargs: Any) -
-            raise ValueError(
-                f"Expected to find at least `{len(categories)}` colors, "
-                f"found `{len(palette)}` for key `{cluster_key}`."
+def _get_palette(
```
as above

giovp replied:
done

giovp commented on .mypy.ini:
```diff
@@ -3,6 +3,7 @@ mypy_path = squidpy
 python_version = 3.8
 plugins = numpy.typing.mypy_plugin
 
+ignore_errors = False
```
nono there are None actually, mypy running fine

giovp commented on squidpy/pl/_spatial_utils.py:
```diff
@@ -0,0 +1,897 @@
+    TYPE_CHECKING,
+)
+from functools import partial
+from collections import namedtuple
```
yes indeed, found out some time ago and it's much nicer

giovp commented on squidpy/pl/_spatial.py:
```diff
@@ -0,0 +1,312 @@
+        _lib_count = first if library_first else second
+        value_to_plot = second if library_first else first
+
+        _size = spatial_params.size[_lib_count]
```
good point, maybe for another time

giovp commented on squidpy/pl/_spatial.py:
```diff
@@ -0,0 +1,764 @@
+    edges_color: Union[str, Sequence[float], Sequence[str]] = "grey",
+    neighbors_key: Optional[str | None] = None,
+    edges_kwargs: Mapping[str, Any] = MappingProxyType({}),
+) -> Any:
```
Done

giovp commented:
ok I think this is now in good shape. 

TODO:
- [ ] add alpha on cmap (see summary above)
- [ ] check documentation
- [ ] add tests (ideas how to proceed @michalk8 ? am thinking to test single functions with dummy data and then couple of instances for main `spatial_` functions. Also, will need to cache couple of datasets).
- [ ] make examples and tutorials

@AnnaChristina @hspitzer @ivirshup I'd really appreciate your help here. Would be great if you can try out API on couple of use cases and provide general feedback on usability+docs (can be improved I think a bit).

You can start with these couple of code snippets

### Test points

<details>
<summary>Details</summary>

```python
import squidpy as sq
from squidpy.pl._spatial import spatial_point, spatial_shape, spatial_segment

adata_seqfish1 = sq.datasets.seqfish()
sq.gr.spatial_neighbors(adata_seqfish1)
adata_seqfish2 = adata_seqfish1.copy()
adata_seqfish = ad.concat(
    {"seqfish1": adata_seqfish1, "seqfish2": adata_seqfish2},
    label="batch_key",
    uns_merge="unique",
    pairwise=True,
)

spatial_point(
    adata_seqfish,
    library_key="batch_key",
    cmap="inferno",
    edges=True,
    edges_width=2,
    size=[30, 50],
    color=["Wnt8a", "celltype_mapped_refined"],
    outline=True,
    outline_width=(0.2, 0.2),
    crop_coord=[[1, 2, 1, 2], [1, 2, 1, 2]],
    scalebar_dx=2.0,
    scalebar_kwargs={"scale_loc": "bottom", "location": "lower right"},
    #     vmin=4,
    vmax=5,
    dpi=100,
)
```

</details>

## Test shapes

<details>
<summary>Details</summary>

```python
import squidpy as sq
from squidpy.pl._spatial import spatial_point, spatial_shape, spatial_segment

adata1 = sq.datasets.visium_hne_adata()
sq.gr.spatial_neighbors(adata1)
adata2 = adata1.copy()
adata2.uns["spatial"] = {}
adata2.uns["spatial"]["V2_Adult_Mouse_Brain"] = adata1.uns["spatial"][
    "V1_Adult_Mouse_Brain"
]

adata_temp = ad.concat(
    {"V1_Adult_Mouse_Brain": adata1, "V2_Adult_Mouse_Brain": adata2},
    label="batch_key",
    uns_merge="unique",
    pairwise=True,
)

spatial_shape(
    adata_temp,
    cmap="inferno",
    shape="square",
    size=[1, 1.25],
    color=["Sox17", "cluster"],
    edges=True,
    edges_width=5,
    title=None,
    outline=True,
    outline_width=(0.05, 0.05),
    crop_coord=(1500, 2800, 1500, 2800),
    scalebar_dx=2.0,
    scalebar_kwargs={"scale_loc": "bottom", "location": "lower right"},
)
```

</details>

## Test segment

<details>
<summary>Details</summary>

```python
import squidpy as sq
from squidpy.pl._spatial import spatial_point, spatial_shape, spatial_segment

adata = sq.datasets.mibitof()
spatial_segment(
    adata,
    color=["Cluster", "cell_size"],
    groups=["Fibroblast", "Endothelial"],
    na_color=(0, 0, 0, 0),
    library_key="library_id",
    cell_id_key="cell_id",
    img=True,
    seg=True,
    seg_boundaries=True,
    seg_erosionpx=15,
    figsize=(5, 5),
    cmap="magma",
    vmin=500,
)

spatial_segment(
    adata,
    color="Cluster",
    #     groups=["Fibroblast", "Endothelial"],
    #     img_alpha=0.5,
    na_color=(0, 0, 0, 0),
    library_key="library_id",
    cell_id_key="cell_id",
    img=True,
    seg=True,
    seg_boundaries=True,
    seg_erosionpx=10,
    figsize=(5, 5),
)
```

</details>

We'd want to cut new release before easter and this is the biggest addition (together with spin off of napari) so would require some thorough beta testing. Thanks all 🙏

hspitzer commented:
Thanks for this great work @giovp !
I played around with the plotting functions a bit and in general they work great!
I think I found one bug:
Plotting only one library_id with `spatial_shape` does not work. 
The following code plots all library_ids, even though I specify 'point8'
```python
adata_mibitof = sq.datasets.mibitof()
sq.pl.spatial_shape(
    adata_mibitof,
    library_key='library_id', 
    library_id=['point8'],
    color=["Cluster", "cell_size"],
    ncols=2
    )
```
The equivalent `spatial_segment` call only plots one library_id as expected:
```python
sq.pl.spatial_segment(
    adata_mibitof,
    library_key='library_id', 
    library_id=['point8'],
    cell_id_key='cell_id',
    color=["Cluster", "cell_size"],
    ncols=2
    )
```
Btw, `spatial_point` on this example fails with a slightly cryptic error message: 
```
# this fails
spatial_point(adata_mibitof, library_key='library_id')
```
```
File ~/projects/spatial_scanpy/squidpy/squidpy/pl/_spatial_utils.py:233, in <listcomp>(.0)
    ]()[231]()[ if scalefactors is not None:
    ]()[232]()[     if scale_factor is None:  # get intersection of scale_factor and match to img_res_key
--> ]()[233]()[         scale_factor_key = [i for i in scalefactors if img_res_key in i]
    ]()[234]()[         if not len(scale_factor_key):
    ]()[235]()[             raise ValueError(f"No `scale_factor` found that could match `img_res_key`: {img_res_key}.")

TypeError: 'in <string>' requires string as left operand, not NoneType]()
```

</details>

A couple of further remarks:
- I was not too sure why you have three different interfaces to `spatial_plot`. For me, it was a bit confusing when to use `spatial_point` and when to use `spatial_shape`. The main difference between them seems to be that `spatial_point` does not plot the image, and shapes are restricted to circles. What is the argument for having two function here? Could you not set the default such that if no image is found no image is plotted? 
I understand the need for a separate `spatial_segment` function though, that makes the interface a lot easier. 
- Would it be possible to have the library_id in the default title as well? When plotting multiple library ids, it can get confusing otherwise
- What exactly do `seg_boundaries` and `set_erosionpx` do? I did not really understand the doctoring, and don't quite understand the potting behaviour with different values of `seg_erosionpx`
- could you allow both lists + strings as input to library_id (or change name to library_ids, to indicate that list input is expected)
- subfigures in `spatial_segment` plot above have slightly different sizes. Is this easily fixable?

slightly off topic, but I noticed that if we have an adata object combining multiple library_ids, the spatial graph does not take into account and is computed using all spatial coordinates. This is not always desirable, and its easy enough to iterate over each library id idenpendently, and concatenate the results again, but I was wondering if maybe `sq.gr.spatial_neighbors` should have an option to compute the graph for each library idependently

giovp commented:
thanks for checking this out @hspitzer , I'll go through point by point

> The following code plots all library_ids, even though I specify 'point8'

bug, fixed it, thanks!

> Btw, spatial_point on this example fails with a slightly cryptic error message:

another bug, fixed it, thanks!

> The main difference between them seems to be that spatial_point does not plot the image, and shapes are restricted to circles.

so `spatial_points` does not indeed plot the image, but simply a scatterplot. `spatial_shape` instead plots shapes, that can be circles, hexagons or squares.
The difference to have this plot separated is that the `spatial_shape` can control the size of the spots/observations (whether they are circles, squares or hexagons) via the `scale_factor`. Therefore, the shapes are correctly placed where the capture area is (for visium, dbit-seq and others). I thought to keep them separately because of this reason. I could also plot the image for the `spatial_points` but then we don't really have a way to infer from data what should be a good size for the scatterplot. Do you think this is an unnecessary concern? happy to hear thoughts on this

> Would it be possible to have the library_id in the default title as well? When plotting multiple library ids, it can get confusing otherwise

this makes sense, to be added

> What exactly do seg_boundaries and set_erosionpx do? I did not really understand the doctoring, and don't quite understand the potting behaviour with different values of seg_erosionpx

They can plot segmentation makss which have erosion outwards meaning that only the "boundaries" of the segmentation makss are plotted, I have an example above where to show this but also here https://github.com/theislab/squidpy/issues/427#issuecomment-1032990172 . The `seg_boundaries` instead plot a "boundary" in black on the segmentation masks, similar to the `outline` argument.

> could you allow both lists + strings as input to library_id (or change name to library_ids, to indicate that list input is expected)

yes indeed, it was a bug, should be fixed

> subfigures in spatial_segment plot above have slightly different sizes. Is this easily fixable?

do you think between continuos and categorical annotations? I noticed it as well but I must say I couldn't ifx it. I think scanpy has the same.

> but I noticed that if we have an adata object combining multiple library_ids, the spatial graph does not take into account and is computed using all spatial coordinates. This is not always desirable, and its easy enough to iterate over each library id idenpendently, and concatenate the results again, but I was wondering if maybe sq.gr.spatial_neighbors should have an option to compute the graph for each library idependently

This is indeed a major problem and something the propagates on all the other functions in the library. The first step would be to allo spatial_neighbors to accept a `batch_key` or `library_key` argument so that it computes the neighbor graph per slide. After that, we'd have to do the same for the rest of the `sq.gr` functions and add tests. Not sure how much work it is but I think it'd be great to have it for next release cause it's really a bottleneck and makes it very unintuitive to users. Any help is very much appreciated.

hspitzer commented:
> > The main difference between them seems to be that spatial_point does not plot the image, and shapes are restricted to circles.
> 
> so `spatial_points` does not indeed plot the image, but simply a scatterplot. `spatial_shape` instead plots shapes, that can be circles, hexagons or squares. The difference to have this plot separated is that the `spatial_shape` can control the size of the spots/observations (whether they are circles, squares or hexagons) via the `scale_factor`. Therefore, the shapes are correctly placed where the capture area is (for visium, dbit-seq and others). I thought to keep them separately because of this reason. I could also plot the image for the `spatial_points` but then we don't really have a way to infer from data what should be a good size for the scatterplot. Do you think this is an unnecessary concern? happy to hear thoughts on this

Just to clarify: Internally both functions use a scatterplot, right? 
Ok, so the reason for separating them is that we don't always have scale information (in fact we only have scale information from the data when an image is stored in `uns['spatial']`. So you think the user might be confused about what the size of the dots/shapes means when we mix data that has a scale factor and data that does not. 
I can see this point, but personally I'm not sure there needs to be a separation; if the data does not have intrinsic scale information, why should the user assume that you plot any? 
Of course separating the image plotting from the simple scatter plot also makes the API easier, which might be a good argument in itself. 
In any case, I would maybe rename `spatial_point` to `spatial_scatter` to match the scanpy API. And then either have `spatial_scatter` display the image if there is one + scale dots/shapes appropriately, or have another function `spatial_image` (or similar), for plotting shapes + images. 
This way it would be a bit more intuitive (at least for me) which function does what.


> > What exactly do seg_boundaries and set_erosionpx do? I did not really understand the doctoring, and don't quite understand the potting behaviour with different values of seg_erosionpx
> 
> They can plot segmentation makss which have erosion outwards meaning that only the "boundaries" of the segmentation makss are plotted, I have an example above where to show this but also here [#427 (comment)](https://github.com/theislab/squidpy/issues/427#issuecomment-1032990172) . The `seg_boundaries` instead plot a "boundary" in black on the segmentation masks, similar to the `outline` argument.

Ok, makes sense! I initially thought erosion would mean that the area you draw would shrink with larger `seg_erosionpx`. But its the other way around. Maybe renaming the arguments would make this clearer?
- `seg_boundaries` -> `seg_outline` (to match the `outline` argument in  `spatial_point`
- `seg_erosionpx` -> `seg_contourpx`
Also, for the `seg_erosionpx` docstring, maybe sth like: "Draw contour of specified width for each segment. If None, fills entire segment." makes it more clear.

> > subfigures in spatial_segment plot above have slightly different sizes. Is this easily fixable?
> 
> do you think between continuos and categorical annotations? I noticed it as well but I must say I couldn't ifx it. I think scanpy has the same.

yes exactly this slight change in size. Ok, if scanpy has the same, then lets just leave it.


> > but I noticed that if we have an adata object combining multiple library_ids, the spatial graph does not take into account and is computed using all spatial coordinates. This is not always desirable, and its easy enough to iterate over each library id idenpendently, and concatenate the results again, but I was wondering if maybe sq.gr.spatial_neighbors should have an option to compute the graph for each library idependently
> 
> This is indeed a major problem and something the propagates on all the other functions in the library. The first step would be to allo spatial_neighbors to accept a `batch_key` or `library_key` argument so that it computes the neighbor graph per slide. After that, we'd have to do the same for the rest of the `sq.gr` functions and add tests. Not sure how much work it is but I think it'd be great to have it for next release cause it's really a bottleneck and makes it very unintuitive to users. Any help is very much appreciated.

Ok! Yes, this would actually be great to have! Not sure if I will have lots of time to work on this, but I could try checking it out.

giovp commented:
> Just to clarify: Internally both functions use a scatterplot, right?

no they don't, `spatial_point` uses `ax.scatter` whereas the shape uses `PatchCollection` to build polygons. 

> Ok, so the reason for separating them is that we don't always have scale information (in fact we only have scale information from the data when an image is stored in uns['spatial']. So you think the user might be confused about what the size of the dots/shapes means when we mix data that has a scale factor and data that does not.
I can see this point, but personally I'm not sure there needs to be a separation; if the data does not have intrinsic scale information, why should the user assume that you plot any?

yeah we don't have both scale and img information sometime. basically if I were to add the `spatial_scatter` function that we discussed, then the `spatial_point` behaviour would be different and user would have to specify
```
spatial_scatter(adata, shape=None, img=False,...)
```
whereas iwth `spatial_point` they don't have to specify that but just use the function out of the box `spatial_point(adata,...)`. This was motivation for separating spatial_scatter in spatial_point and spatial_shapes but maybe wrong, which API do you think is better?

> Ok, makes sense! I initially thought erosion would mean that the area you draw would shrink with larger seg_erosionpx. But its the other way around. Maybe renaming the arguments would make this clearer?

seg_boundaries -> seg_outline (to match the outline argument in spatial_point
seg_erosionpx -> seg_contourpx
Also, for the seg_erosionpx docstring, maybe sth like: "Draw contour of specified width for each segment. If None, fills entire segment." makes it more clear.

sounds good, done.

giovp commented:
ok, decided to go for `spatial_scatter`, thanks @hspitzer for discussion. Now to plot a scatterplot without image just need to do`sq.pl.spatial_scatter(adata, shape=None)` .

hspitzer commented:
Ok, cool!
What will happen if there is no image data (eg seqfish)? Will the function still plot
Why are you using the name `shape` to control if an image is shown? Wasn't this parameter used to control the size of the dots?

giovp commented:
> What will happen if there is no image data (eg seqfish)? Will the function still plot
it will just plot the scatter plot

> Why are you using the name shape to control if an image is shown? Wasn't this parameter used to control the size of the dots?

no this argument set whether a shape (circle, square, hex) will be used. If None, a scatterplot will be used (and no image is plotted)

AnnaChristina commented:
pasting my comments here, please ignore them if this is intended behaviour:

### 1. mibitof data

```
sq.pl.spatial_scatter(
    adata_mibitof,
    color="Cluster",
    img=False
)
```
![image](https://user-images.githubusercontent.com/44966624/161041009-6911b637-c066-4e00-8c36-77076a2f68d4.png)

plots the same image three times without the underlying image, I would expect that it shows all present images side by side
```
sq.pl.spatial_scatter(
    adata_mibitof,
    color="Cluster",
    img=True
)
```
![image](https://user-images.githubusercontent.com/44966624/161041106-2087e5ad-2dbc-4157-9956-e164f5770a70.png)

correctly changes the background image, but is not updating the scatter plot. 


### 2. other data
I additionally tested the function on a second dataset where I collected all items needed for this function. General comment would be that we need a good tutorial that guides you through the steps of creating this structure. I think many people would like to use this function and its super nice. 

one think I realized is the following:
![image](https://user-images.githubusercontent.com/44966624/161052437-c85da6e0-0108-4dcf-ad4b-5b5e158fe9e0.png)

if I plot the annotation for my example and the outline for plotting the continuous Area is correctly captured, but for the annotation, the outline disappears if two cells of the same type are super close. 


I will explore further now and move my notebook for these kinds of plots towards the new function and will add my comments below if I come across more questions/hints/issues. 

Overall this looks super cool imho! brings the whole plotting of spatial omics datasets to a whole new level and super excited that you were pushing for this.

ivirshup commented:
> if I plot the annotation for my example and the outline for plotting the continuous Area is correctly captured, but for the annotation, the outline disappears if two cells of the same type are super close.

@giovp, I think this is due to the boundaries being found after mapping cells -> categories, and could be solved by finding boundaries from the cell segmentation directly.

giovp commented:
thanks a lot for testing it @AnnaChristina ! So for mibitof I've fixed the behaviour, should be consistent now. 

for the boundaries, that's interesting cause it never happens with the mibitof data:

<img width="689" alt="image" src="https://user-images.githubusercontent.com/25887487/162404614-faa89585-8953-4ef3-a6c8-7b2fe02134cb.png">

>  I think this is due to the boundaries being found after mapping cells -> categories, and could be solved by finding boundaries from the cell segmentation directly.

@ivirshup I think I get your point, I change the input of `find_boundaries` to original segmentation mask. @AnnaChristina could you try again on that dataset? If still doesn't work, I might have to allow to pass kwargs to find_boundaries https://scikit-image.org/docs/stable/api/skimage.segmentation.html#find-boundaries

giovp commented:
`tox -e check-docs` runs successfully locally....

michalk8 commented on squidpy/pl/_spatial_utils.py:
```diff
@@ -0,0 +1,960 @@
+    hspace: float,
+    wspace: float,
+    ncols: int,
+    figsize: tuple[float, float],
```
Sometimes you use `tuple` or `Tuple`, would unify (low prio).

michalk8 commented on squidpy/_constants/_pkg_constants.py:
```diff
@@ -134,6 +138,63 @@ def spot_diameter(
+            return {i: list(adata.uns[spatial_key][i][sub_key].keys()) for i in library_id}
+
+        @classmethod
+        def _haystack(
```
Please find a better name for this function

giovp replied:
`_sort_haystack` ?

michalk8 commented on squidpy/_constants/_pkg_constants.py:
```diff
@@ -134,6 +138,63 @@ def spot_diameter(
+            spatial_key: str,
+            library_id: Sequence[str] | str | None = None,
+            sub_key: Optional[str] = None,
+        ) -> Sequence[str]:
```
Sequence[str] | None

michalk8 commented on squidpy/_constants/_pkg_constants.py:
```diff
@@ -134,6 +138,63 @@ def spot_diameter(
+                        f"Please specify one from: `{sorted(library_id)}`."
+                    )
+                return library_id[0]
+            if len(library_id) > 0:
```
Library id can be None here, consider using:
```python
if return_all or library_id is None:
    return library_id
if len(library_id) != 0:
    raise ValueError(...)
return library_id[0]
```

michalk8 commented on squidpy/_constants/_pkg_constants.py:
```diff
@@ -134,6 +138,63 @@ def spot_diameter(
+        ) -> Sequence[str]:
+            if spatial_key not in adata.uns:
+                raise KeyError(f"Spatial key {spatial_key!r} not found in `adata.uns`.")
+            haystack = list(adata.uns[spatial_key].keys())
```
There are some unnecessary calls to `.keys` (usually, it should suffice that the object implement `__getitem__` only); won't mention elsewhere

giovp replied:
I see the point but I think it's more understandable

michalk8 commented on squidpy/_constants/_pkg_constants.py:
```diff
@@ -134,6 +138,63 @@ def spot_diameter(
+                            f"`{sub_key}` not found in `adata.uns[{spatial_key!r}]['library_id'].keys()` "
+                            f"with following `library_id`: {library_id}."
+                        )
+                    return library_id
```
Unnecessary return (let if fall through)

michalk8 commented on squidpy/_docs.py:
```diff
@@ -174,6 +174,243 @@ def decorator2(obj: Any) -> Any:
+    Select one or some of the unique `library_id` that constitute the AnnData to plot."""
+_library_key = """\
+library_key
+    If multiple `library_id`, column in :attr:`anndata.AnnData.obs` which stores mapping between `library_id` and obs"""
```
Missing `.` at the end.

michalk8 commented on squidpy/_docs.py:
```diff
@@ -174,6 +174,243 @@ def decorator2(obj: Any) -> Any:
+
+If your anndata object has a `"spatial"` entry in `.uns`, the `img_key`, `seg_key`
+and `library_id` parameters to find values for `img`, `seg`, `scale_factor`,
+and `spot_size` arguments. Alternatively, these values be passed directly.
```
`can be`

michalk8 commented on squidpy/pl/_color_utils.py:
```diff
@@ -0,0 +1,53 @@
+def _get_palette(
+    adata: AnnData, cluster_key: str, categories: Sequence[Any], palette: Palette_t = None
+) -> Mapping[str, str] | None:
+    if palette is None:
```
Seems also natural to handle `palette` as a `Sequence[str | NDArrayA]`, but low prio

giovp replied:
ok will leave it then

michalk8 commented on squidpy/pl/_spatial.py:
```diff
@@ -0,0 +1,415 @@
+    )
+
+    for count, (first, second) in enumerate(itertools.product(*fig_params.iter_panels)):
+        _lib_count: int = first if library_first else second
```
Why not name `first/second` as `_lib_count/value_to_plot` and then
```python
if not library_first:
    _lib_count, value_to_plot = value_to_plot, _lib_count
```

michalk8 commented on squidpy/pl/_spatial.py:
```diff
@@ -0,0 +1,415 @@
+        **kwargs,
+    )
+
+    _subset: Callable[[AnnData, str | None, str | None], AnnData] = (
```
Tbh don't like this that much, saves you almost nothing (+ have to ignore types).
Would just later use `adata_sub = adata if library_key is None else _subs(adata, library_key=library_key, library_id=_id)` (low prio)

giovp replied:
I use it twice though thought it's nicer

michalk8 commented on squidpy/pl/_spatial.py:
```diff
@@ -0,0 +1,415 @@
+        adata,
+        shape=shape,
+        img=img,
+        seg_key=None,
```
Just a question: does this really need to set just 1 argument?

giovp replied:
`shape` and `img` have also different defaults.

michalk8 commented on squidpy/pl/_spatial_utils.py:
```diff
@@ -0,0 +1,960 @@
+    img_res_key: str | None = None,
+    img_channel: int | None = None,
+    img_cmap: Colormap | str | None = None,
+) -> Union[Sequence[NDArrayA], Tuple[None, ...]]:
```
List[None]

giovp replied:
mypy throws error cause it's a generator

michalk8 commented on squidpy/pl/_spatial_utils.py:
```diff
@@ -0,0 +1,960 @@
+    library_key: str | None = None,
+    seg: _SeqArray | bool = None,
+    seg_key: str | None = None,
+) -> Tuple[Sequence[NDArrayA], Sequence[NDArrayA]] | Tuple[Tuple[None, ...], Tuple[None, ...]]:
```
Would simplify to (based on code):
```python
Tuple[Sequence[NDArrayA] | Sequence[None], Sequence[NDArrayA]]
```

giovp replied:
mypy throws errors

michalk8 commented on squidpy/pl/_spatial_utils.py:
```diff
@@ -0,0 +1,960 @@
+            img_cmap=img_cmap,
+        )
+    else:
+        _img = tuple(None for _ in library_id)
```
Or `(None,) * len(library_id)`

michalk8 commented on squidpy/pl/_spatial_utils.py:
```diff
@@ -0,0 +1,960 @@
+    library_id = _get_library_id(
+        adata=adata, shape=shape, spatial_key=spatial_key, library_id=library_id, library_key=library_key
+    )
+    if len(library_id) > 1 and library_key is None:
```
Can this really happen? If yes, I'd say `_get_library_id` should check this (if it doesn't break anything).

michalk8 reviewed (changes requested):
Looks nice, minor issues, will give it a go later today.

giovp commented on tests/plotting/test_spatial_static.py:
```diff
@@ -0,0 +1,204 @@
+
+
+class TestSpatialStaticUtils:
+    def _create_anndata(self, shape, library_id, library_key):
```
done

giovp commented on squidpy/pl/_spatial_utils.py:
```diff
@@ -0,0 +1,960 @@
+
+def _subs(adata: AnnData, library_key: str | None = None, library_id: str | None = None) -> AnnData:
+    try:
+        return adata[adata.obs[library_key] == library_id]
```
done

giovp commented on squidpy/pl/_spatial_utils.py:
```diff
@@ -0,0 +1,960 @@
+    legend_fontweight: int | _FontWeight = "bold",
+    legend_loc: str = "right margin",
+    legend_fontoutline: Optional[int] = None,
+    na_color: str | Tuple[float, ...] = (0.0, 0.0, 0.0, 0.0),
```
it returned mypy errors that's why I used that

giovp commented on squidpy/pl/_spatial_utils.py:
```diff
@@ -0,0 +1,960 @@
+
+_AvailShapes = Literal["circle", "square", "hex"]
+Palette_t = Optional[Union[str, ListedColormap]]
+_Normalize = Union[Normalize, Sequence[Normalize], None]
```
>  I'd prefer no None in the aliases

you mean replace it with `Optional` ?

michalk8 replied:
No, I meant remove `Optional`/`Union[..., None]` altogether from the alias and making it explicit in the function signature.

giovp replied:
done

ivirshup commented on squidpy/pl/_spatial_utils.py:
```diff
@@ -0,0 +1,960 @@
+
+    if norm is not None:
+        if (vmin is not None) or (vmax is not None) or (vcenter is not None):
+            raise ValueError("Passing both norm and vmin/vmax/vcenter is not allowed.")
```
I wouldn't use `logg.warning` any reasonable logging system will log normal warnings

giovp replied:
think it's fine for hte use case

michalk8 commented on squidpy/pl/_spatial_utils.py:
```diff
@@ -0,0 +1,960 @@
+    except KeyError:
+        raise KeyError(
+            f"Cannot subset adata. Either `library_key: {library_key}` or `library_id: {library_id}` is invalid."
+        )
```
I meant use `raikse KeyError(f"Cannot ...") from None`, that should be fine.

giovp commented on squidpy/pl/_spatial_utils.py:
```diff
@@ -0,0 +1,960 @@
+    cmap = copy(get_cmap(cmap))
+    cmap.set_bad("lightgray" if na_color is None else na_color)
+
+    vmin = kwargs.pop("vmin", None)
```
they are already a lot, ok for now to keep in kwargs imho

giovp commented on squidpy/pl/_utils.py:
```diff
@@ -621,3 +621,18 @@ def _dendrogram(data: NDArrayA, method: str, **kwargs: Any) -> tuple[list[int],
+
+def _assert_value_in_obs(adata: AnnData, key: str, val: Sequence[Any] | Any) -> None:
+    if key not in adata.obs:
+        raise KeyError(f"Cluster key `{key}` not found in `adata.obs`.")
```
great point

giovp commented on tests/plotting/test_spatial_static.py:
```diff
@@ -0,0 +1,204 @@
+C_KEY = "Cluster"
+
+
+sc.pl.set_rcParams_defaults()
```
done

giovp commented on tests/plotting/test_spatial_static.py:
```diff
@@ -0,0 +1,204 @@
+            library_id=library_id,
+            library_key=library_key,
+        )
+        if shape is None:
```
the pytest.raise check is different though

giovp commented on squidpy/_docs.py:
```diff
@@ -174,6 +174,243 @@ def decorator2(obj: Any) -> Any:
+    Whether to plot scatter plot of points or regular polygons."""
+_color = """\
+color
+    Which features to plot from :class:`anndata.AnnData`."""
```
changed, copied it over from scanpy

giovp commented on squidpy/_docs.py:
```diff
@@ -174,6 +174,243 @@ def decorator2(obj: Any) -> Any:
+Nothing, just plots and optionally saves the plot.
+"""
+
+_plotting_general_summary = """\
```
is addes `%(plotting_general_summary)s` to `spatial_plot`, what do you mean by extended summary?

giovp commented on squidpy/pl/_spatial.py:
```diff
@@ -0,0 +1,415 @@
+
+@d.dedent
+@inject_docs(key=Key.obsp.spatial_conn())
+def spatial_plot(
```
cause this is a general function that users should maybe not use, like "embedding" (at least this is the way I see it). I'm even considering to make it hidden actually. What do you think?

michalk8 replied:
As
```python
@d.get_extended_summary(base='foo')
def spatial_plot(...):
    ...

@d.dedent
def spatial_scatter():
     """
     %(foo.extended_summary)s  # not sure if correct key 
     """
```

giovp commented on squidpy/pl/_spatial_utils.py:
```diff
@@ -0,0 +1,960 @@
+    layer: str | None = None,
+    palette: Palette_t = None,
+) -> ColorParams:
+    alpha = 1.0 if alpha is None else alpha
```
good catch
